types/react/test/hooks.tsx

```diff
@@ -299,7 +321,7 @@ function useExperimentalHooks() {
     // $ExpectType boolean
     const deferredToggle = React.useDeferredValue(toggle);
 
-    const [func] = React.useState(() => () => 0);
+    const [func] = React.useState<() => number>(() => () => 0);
```

Comment:
This is the only piece I'm concerned about. Somehow this change narrowed `() => () => 0` to an `S` of `() => 0` instead of `() => number`.

Would love any thoughts on how we can avoid this specific issue, since it might cause breaks, even though it's technically correct.

Reply:
I don't know how to fix this particular issue but requiring more type annotations is pretty unpopular. Could you split the PR into multiple commits so that it's clear what the behavior of the new tests was before and what it is with the proposed changes? That makes it easier to gauge if this change should land.

Reply:
Apologies for the late reply @eps1lon! I tried to break the commits up a bit better, but let me know if there's something better I can do here!

1. Introduces the new tests & assert the current behavior (these tests have runtime breaks): https://github.com/DefinitelyTyped/DefinitelyTyped/pull/63609/commits/dabda4b8b102a960e4f2a0853e97c319a0bd596d
2. Corrects the tests to assert the expected, correct behavior. https://github.com/DefinitelyTyped/DefinitelyTyped/pull/63609/commits/879edfc5560adb854b00ab24c378c19d62e442ae
3. The actual typing changes: https://github.com/DefinitelyTyped/DefinitelyTyped/pull/63609/commits/6f626cbbcd69448b4d40be2e6003e36b7c954e09
4. The unfortunate, extra type annotations, as noted in this thread: https://github.com/DefinitelyTyped/DefinitelyTyped/pull/63609/commits/b3e3f08892e900f147008d39bf480bd4c5e3f74a

Reply:
Alternatively, you could also write `const [func] = useState(() => (): number => 0)` which would no longer narrow the return type to const.

It also doesn't require additional typing if the passed function is not contextually typed:
```ts
const getNumber = () => 0
const [getNumberState] = useState(() => getNumber)
```